src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+          // be retried
+          console.warn(e);
+          return subscriber.error(e);
+        });
+    });
+    return resp$;
+  }
+
+  private postMessageWithAttachments(message: Message) {
+    const { attachments, ...messageWithoutAttachments } = message;
+
+    let httpContentList = [];
+    return Observable.create(subscriber => {
+      return Observable.from(attachments || [])
+        .flatMap((media: Media) =>
+          Observable.ajax({
+            method: "GET",
+            url: media.contentUrl,
+            responseType: 'arraybuffer'
+          })
+            .do(ajaxResponse => {
+              let buffer = new Buffer(ajaxResponse.response);
+              let stream = new BFSE.SubscribableStream();
+              stream.write(buffer);
+              let httpContent = new BFSE.HttpContent({ type: media.contentType, contentLength: buffer.length }, stream);
+              httpContentList.push(httpContent);
+            }))
+        .count()
+        .flatMap(_ => {
+          let url = `/v3/directline/conversations/${this.conversationId}/users/${messageWithoutAttachments.from.id}/upload`;
+          let request = BFSE.StreamingRequest.create('PUT', url);
+          var activityStream = new BFSE.SubscribableStream();
+          activityStream.write(JSON.stringify(messageWithoutAttachments), 'utf-8');
+          request.addStream(new BFSE.HttpContent({ type: "application/vnd.microsoft.activity", contentLength: activityStream.length }, activityStream));
+          httpContentList.forEach(e => request.addStream(e));
+          return this.streamConnection.send(request);
+        })
+        .do(resp => {
+          if (resp.streams && resp.streams.length !== 1) {
+            subscriber.error("Invalid stream count " + resp.streams.length);
+          } else {
+            resp.streams[0].readAsJson()
+              .then(json => {
+                subscriber.next(json['Id'])
+              })
+          }
+        })
+        .subscribe(_ => _) // force execution
+    })
+  }
+
+  private disconnectionHandler(e: any) {
+    if (this.connectionStatus$.value === ConnectionStatus.Connecting ||
+        this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    if (this.token === null) {
+      this.activitySubscriber.error("Token unavailable");
+      return;
+    }
+
+    this.connectionStatus$.next(ConnectionStatus.Connecting);
+    this.retryCount--;
+    if (this.retryCount > 0) {
+      setTimeout(this.streamingWebSocketActivity$.bind(this), this.getRetryDelay());
+    } else {
+      console.warn("Exhausted retries");
+      this.activitySubscriber.error(e);
+    }
+  }
+
+  private streamingWebSocketActivity$(): Observable<Activity> {
+    if (this.activitySubscriber) {
+      this.theStreamHandler.setSubscriber(this.activitySubscriber);
+      this.connectAsync().then(_ => _);
+    } else {
+      return Observable.create(async (subscriber: Subscriber<Activity>) => {
```

Comment:
In line 118, we assign this `Observable.create` to `this.activity$`.

But when Web Socket failed, line 278, we just call this function again, but the `this.activity$` is not updated. Is that correct?

It feels like the retry logic is difficult to read/debug and jump around using `setTimeout`. I think we did a bad job in DLJS and should do better here.